File: src/adapters/driver/rtmp/rtmpServer.go
package rtmp

import (
	"fmt"
	"io"
	"log"
	"net"
	"strconv"

	"Theatrum/adapters/driver/ports"
	"Theatrum/adapters/driver/rtmp/config"
	rtmphandler "Theatrum/adapters/driver/rtmp/handlers"
	stream "Theatrum/adapters/driver/rtmp/management"
	"Theatrum/domain/services"

	"github.com/yutopp/go-rtmp"
)

// RtmpServer implements the RtmpPort interface
type RtmpServer struct {
	applicationService *services.ApplicationService
	streamService      *services.StreamService
	rtmpAuthService    *services.RtmpAuthService
	server             *rtmp.Server
	listener           net.Listener
	streamManager      *stream.Manager
}

// Verify interface implementation
var _ ports.RtmpPort = (*RtmpServer)(nil)

func NewRtmpServer(applicationService *services.ApplicationService, streamService *services.StreamService, rtmpAuthService *services.RtmpAuthService) ports.RtmpPort {
	return &RtmpServer{
		applicationService: applicationService,
		streamService:      streamService,
		rtmpAuthService:    rtmpAuthService,
		streamManager:      stream.NewManager(),
	}
}

func (s *RtmpServer) StartRtmpServer() error {
	log.Printf("=== RTMP SERVER ===")

	port := s.applicationService.GetServer().RTMPPort
	addr := ":" + strconv.Itoa(port)

	// Create RTMP server
	s.server = rtmp.NewServer(&rtmp.ServerConfig{
		OnConnect: func(conn net.Conn) (io.ReadWriteCloser, *rtmp.ConnConfig) {
			return conn, &rtmp.ConnConfig{
				Handler: rtmphandler.NewHandler(s.rtmpAuthService, s.streamManager, s.getConfig()),
			}
		},
	})

	// Start listening
	ln, err := net.Listen("tcp", addr)
	if err != nil {
		return fmt.Errorf("failed to start RTMP server: %w", err)
	}
	s.listener = ln

	log.Printf("RTMP server listening on %s", addr)

	// Serve synchronously (blocking)
	if err := s.server.Serve(ln); err != nil {
		log.Printf("RTMP server error: %v", err)
		return err
	}

	return nil
}

func (s *RtmpServer) ShutdownRtmpServer() error {
	log.Printf("Shutting down RTMP server...")

	// Close the listener first
	if s.listener != nil {
		if err := s.listener.Close(); err != nil {
			log.Printf("Error closing RTMP listener: %v", err)
		}
	}

	// Stop all active streams
	if s.streamManager != nil {
		activeStreams := s.streamManager.GetActiveStreams()
		for _, inputPath := range activeStreams {
			log.Printf("Stopping stream for: %s", inputPath)
			// Note: The stream manager doesn't have a GetStream method
			// The streams will be cleaned up automatically when the server shuts down
		}
	}

	log.Printf("RTMP server shutdown complete")
	return nil
}

func (s *RtmpServer) GetActiveStreams() []string {
	if s.streamManager != nil {
		return s.streamManager.GetActiveStreams()
	}
	return []string{}
}

// getConfig returns a configuration object for the RTMP server from domain config
func (s *RtmpServer) getConfig() config.Config {
	rtmpConfig := s.applicationService.GetServer().RTMP
	return config.Config{
		ReconnectDelay: rtmpConfig.ReconnectDelay,
		CleanupDelay:   rtmpConfig.CleanupDelay,
	}
}

File: src/main/java/com/fjourdren/theatrum/infrastructure/adapter/in/rtmp/RtmpServer.java
package com.fjourdren.theatrum.infrastructure.adapter.in.rtmp;

import com.fjourdren.theatrum.application.port.in.AuthorizePublishUseCase;
import com.fjourdren.theatrum.application.port.in.LiveStreamVarsUseCase;
import com.fjourdren.theatrum.application.port.in.PathTemplateUseCase;
import com.fjourdren.theatrum.application.port.in.ResolveChannelUseCase;
import com.fjourdren.theatrum.domain.model.AppPaths;
import com.fjourdren.theatrum.domain.model.Rtmp;
import com.fjourdren.theatrum.infrastructure.adapter.in.rtmp.config.RtmpConfig;
import com.fjourdren.theatrum.infrastructure.adapter.in.rtmp.handlers.TheatrumRtmpHandler;
import com.fjourdren.theatrum.infrastructure.adapter.in.rtmp.management.StreamManager;
import com.fjourdren.theatrum.infrastructure.adapter.in.rtmp.protocol.RtmpConnection;
import com.fjourdren.theatrum.infrastructure.adapter.out.metrics.Metrics;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/** Accepts RTMP publishers and serves each one on its own virtual thread. */
@Slf4j
@Component
@RequiredArgsConstructor
public class RtmpServer implements RtmpLifecycle {

    private final ResolveChannelUseCase applicationService;
    private final AuthorizePublishUseCase rtmpAuthService;
    private final PathTemplateUseCase templateService;
    private final LiveStreamVarsUseCase registry;
    private final StreamManager streamManager;
    private final Metrics metrics;
    private final AppPaths appPaths;

    private volatile ServerSocket listener;
    private volatile ExecutorService executor;
    private volatile boolean running;

    /**
     * Binds the RTMP port and returns; the accept loop runs on its own virtual thread so Spring
     * startup is not blocked (the Go server served synchronously).
     */
    @Override
    public void startRtmpServer() {
        int port = applicationService.getServer().rtmpPort();
        log.info("=== RTMP SERVER (port {}) ===", port);

        try {
            listener = new ServerSocket(port);
        } catch (IOException e) {
            throw new UncheckedIOException("failed to start RTMP server", e);
        }

        running = true;
        executor = Executors.newVirtualThreadPerTaskExecutor();
        log.info("RTMP server listening on :{}", port);
        executor.submit(this::acceptLoop);
    }

    @Override
    public void shutdownRtmpServer() {
        log.info("Shutting down RTMP server");
        running = false;

        if (listener != null) {
            try {
                listener.close();
            } catch (IOException e) {
                log.warn("Error closing RTMP listener: {}", e.getMessage());
            }
        }
        // Active streams stop on their own once their FFmpeg process exits, as in the Go server.
        for (String inputPath : getActiveStreams()) {
            log.info("Stopping stream for: {}", inputPath);
        }
        if (executor != null) {
            executor.shutdownNow();
        }

        log.info("RTMP server shutdown complete");
    }

    @Override
    public List<String> getActiveStreams() {
        return streamManager.getActiveStreams();
    }

    private void acceptLoop() {
        while (running) {
            Socket socket;
            try {
                socket = listener.accept();
            } catch (IOException e) {
                if (running) {
                    log.warn("RTMP server error: {}", e.getMessage());
                }
                return; // the listener is closed: shutting down, or unusable either way
            }
            executor.submit(() -> serve(socket));
        }
    }

    private void serve(Socket socket) {
        try {
            new RtmpConnection(socket, newHandler()).serve();
        } catch (IOException e) {
            log.warn("Failed to serve RTMP connection: {}", e.getMessage());
        }
    }

    /** One handler per connection: it carries that connection's authentication and stream state. */
    private TheatrumRtmpHandler newHandler() {
        Rtmp rtmp = applicationService.getServer().rtmp();
        return new TheatrumRtmpHandler(rtmpAuthService, templateService, registry, streamManager,
                new RtmpConfig(rtmp.reconnectDelay(), rtmp.cleanupDelay()), metrics, appPaths);
    }
}
